### PR TITLE
#Refocus redis lens cache

### PR DESCRIPTION
**Pull request by pallavi2209**
- Created redis client which acts as cache.
- GET /v1/lenses/id : Populate cache on get if cache empty.
- lens db afterDelete hook: delete cache entry
- lens db afterCreate hook: populate cache with new entry
- lens db afterUpdate hook: remove cached entry, cache will be populate on GET /v1/lenses/id

### CODE REVIEW

File: db/model/lens.js [hunk 1]
```diff
@@ -137,6 +127,29 @@ module.exports = function lens(seq, dataTypes) {
...
+        redisCache.del(inst.name);
+      },
+
+      afterCreate(inst /* , opts */) {
+        // require inside hook to remove cyclic dependency.
+        const cleanAndCreateLensJson = require(
+        '../../api/v1/controllers/lenses'
+        ).cleanAndCreateLensJson;
+
+        const lensObj = cleanAndCreateLensJson(inst);
```
Comment (iamigo) [inline db/model/lens.js]: I'd like to avoid having any dependencies *from* db layer *into* api layer--it can get very messy!

**Comment by pallavi2209** [comment]
@iamigo updated the PR

**Comment by shriramshankar** [comment]
looks like I commented on the wrong PR last time. This looks good to me.

**Comment by iamigo** [comment]
@pallavi2209, did you follow up on @shriramshankar's question about socket.io realtime events after yesterday's standup? Confirm events sending OK in a multi-dyno heroku deployment?

**Comment by pallavi2209** [comment]
@iamigo I am not able to deploy the app on heroku since I do not have admin access. I have requested the admin access from @meynet . Once that is done, I will test this.

**Comment by pallavi2209** [comment]
Good that we tested this, I am getting this error on multi-dyno env:
       

>        events.js:141
>        throw er; // Unhandled 'error' event
>        ^
>        ReplyError: ERR max number of clients reached
>        at parseError (/tmp/build_2847f70cc163aab66d84c57af407ec99/salesforce-refocus-47357c4/node_modules/redis-parser/lib/parser.js:163:12)

It works fine on single dyno though. Let me se if I can do the caching using the already existing redis clients.

**Comment by pallavi2209** [comment]
So far what I have discovered is the error goes away when I set WEB_CONCURRENY=1 (makes sense because every process tries to create a redis client and it max out). But even afetr this change, realtime events does not work. I need to dig more into this.

**Comment by pallavi2209** [comment]
@iamigo @shriramshankar rebased the PR correctly, events work fine now. Check review app: [http://refocus-pr-18.herokuapp.com/perspectives/persp-bracket-lens?debug=REALTIME](http://refocus-pr-18.herokuapp.com/perspectives/persp-bracket-lens?debug=REALTIME)

**Comment by shriramshankar** [comment]
great. ok to merge.

**Pull request merged**